### DISCUSSION TITLE
Feature Request: support GitHub Apps to bypass pull request requirements

### DISCUSSION THREAD

**Discussion by ReinAchten-TomTom** [community/community#7986] [Pull Requests]
Next to supporting people and teams, it would be great if GitHub Apps could be supported as well in the [recently introduced feature to bypass pull request requirements](https://github.blog/changelog/2021-11-19-allow-bypassing-required-pull-requests/). The description in the feature introduction in the blog already hints at that:

> For example, if you have an automated process that calls GitHub APIs to make changes in a repository, you may want to permit that automation to make changes without creating a pull request.

Guess that's for now limited to automation that uses a PAT for authentication, whereas for automation purposes it would be better to authenticate as a GitHub App instead.

**Comment by SebastienGllmt** [comment]
Presumably this is planned because it's [mentioned in the docs](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#restrict-who-can-push-to-matching-branches) that this is possible despite the fact that, in my experience, this is not possible yet

**Comment by jhohertz** [comment]
Is it possible the search box in the UI is simply not also searching for the apps, as 'restrict who could push' does?

Being able to specify an app as an actor under this would be extremely beneficial, as we can avoid a lot of less ideal config scenarios by letting the CI through regardless of the rules for people.

**Comment by ReinAchten-TomTom** [comment]
FYI It seems this request has been honored as part of https://github.blog/changelog/2022-05-17-consistently-allow-github-apps-as-exceptions-to-branch-protection-rules/ :tada: 

I haven't checked yet, but the description for the aforementioned feature now includes "or apps":
![image](https://user-images.githubusercontent.com/66829057/168970963-ae0238e1-862e-4089-867f-bcb27be2c13a.png)

**Reply by jhohertz** [reply]
I can add app as an actor now, but it doesn't seem actually allow the bypass so far in testing, throwing up the branch protection rule, claiming a status check that is in state success, is not.

**Reply by jhohertz** [reply]
It's because there's some requirement that this not be used with "require status check" enabled.

**Reply by ReinAchten-TomTom** [reply]
hmm :( I'll also try and find some time to check it out sometime soon and will update here with my findings.

**Reply by ejholmes** [reply]
We've also had problems getting this to actually work. We have a repo configured with required PR's + status checks. Despite that, the GitHub App can't commit directly to `main`:

```python
{"type":"BRANCH_PROTECTION_RULE_VIOLATION","path":["createCommitOnBranch"],"locations":[{"line":3,"column":3}],"message":"protected branch 'main' check failed:
  3 of 3 required status checks are expected. At least 1 approving review is required by reviewers with write access."}
```

**Comment by decyjphr** [comment]
Could we extend this feature to include deploy keys to be exempted from branch protections?

**Comment by lutzwillek-tomtom** [comment]
Hey @ReinAchten-TomTom good news: [As per May 17, 2022](https://github.blog/changelog/2022-05-17-consistently-allow-github-apps-as-exceptions-to-branch-protection-rules/), branch protections that previously only allowed exceptions for people and teams also support GitHub Apps.

For more information: https://github.blog/changelog/2022-05-17-consistently-allow-github-apps-as-exceptions-to-branch-protection-rules/

Four our use case, it is working perfectly fine. Thanks for asking, and thanks for implementing!

**Reply by dataviruset** [reply]
Unfortunately it still doesn't allow GitHub Apps to bypass required Status Checks if you have enabled that :(
